OVA creation elastic 7

DFolchA opened the pull request:
Hello team,

This PR closes #176, I have modified the scripts that generate the OVA to take into account that the installation of the elastic stack is different for elastic 6 and 7.

To make the scripts simpler and easier to understand I have created two libraries that contain the necessary functions to install elastic stack and wazuh and added the config_files directory from which the scripts take the configuration files.

To test this I have connected a 3.9.2 agent to the manager in the ova:
- [x]  wazuh 3.9.3 with elastic 7.2.0
- [x]  wazuh 3.9.2 with elastic 7.1.1 
- [x]  wazuh 3.8.2 with elastic 6.7.1 

Regards, 
Daniel Folch

DFolchA commented on ova/Config_files/elasticsearch.yml:
```diff
@@ -0,0 +1,7 @@
+cluster.name: wazuh-cluster
+node.name: ${HOSTNAME}
```
This variable will be replaced with the environment variable, this information can be found in this page of the [elasticsearch documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/settings.html).

DFolchA commented on ova/Config_files/api_config.json:
```diff
@@ -0,0 +1,22 @@
+        "insecure": "true",
+        "component": "API",
+        "cluster_info": {
+        "manager": "wazuh-manager",
```
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/df2320802601b207c199c01425748724cca9eb65

BraulioV commented on ova/Config_files/ossec.conf:
```diff
@@ -0,0 +1,18 @@
+<ossec_config>
```
Can you change the indentation of this XML file? I mean, when a new block starts, increase the indentation to improve the readability of this file.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Config_files/wazuh-template.json:
```diff
@@ -0,0 +1,1455 @@
+{
```
I think this file isn't necessary. The script is downloading it in this line: https://github.com/wazuh/wazuh-packages/blob/df2320802601b207c199c01425748724cca9eb65/ova/Libraries/elastic_functions.sh#L27

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+    cp -f ${config_files}/filebeat.yml ${etc_filebeat}/filebeat.yml
+
+    sed -i "s/YOUR_ELASTIC_SERVER_IP/localhost/" ${etc_filebeat}/filebeat.yml
+    chmod go+r ${etc_filebeat}/filebeat.yml
```
This line isn't necessary.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+    sed -i "s/YOUR_ELASTIC_SERVER_IP/localhost/" ${etc_filebeat}/filebeat.yml
+    chmod go+r ${etc_filebeat}/filebeat.yml
+    curl -so ${etc_filebeat}/wazuh-template.json https://raw.githubusercontent.com/wazuh/wazuh/v${WAZUH_VERSION}/extensions/elasticsearch/7.x/wazuh-template.json
+    chmod go+r ${etc_filebeat}/wazuh-template.json
```
This line isn't necessary.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+}
+
+install_kibana_app(){
+    if [ "${STATUS_PACKAGES}" == "stable" ]; then
```
Change this line from `if [ "${STATUS_PACKAGES}" == "stable" ]; then` to `if [ "${STATUS_PACKAGES}" = "stable" ]; then`.

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+        sudo -u kibana ${usr_kibana}/bin/kibana-plugin install https://packages.wazuh.com/wazuhapp/wazuhapp-${WAZUH_VERSION}_${ELK_VERSION}.zip
+    fi
+
+    if [ "${STATUS_PACKAGES}" == "unstable" ]; then
```
Change this line from `if [ "${STATUS_PACKAGES}" = "unstable" ]; then` to `if [ "${STATUS_PACKAGES}" = "unstable" ]; then`.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+    systemctl enable kibana.service
+    systemctl start kibana
+
+    while true; do
```
This loop may be simplified if you use an `until ...; do stuff; done` structure. Here you can see an example: https://github.com/wazuh/wazuh-packages/blob/df2320802601b207c199c01425748724cca9eb65/ova/Libraries/elastic_functions.sh#L163-L166

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/63e8681803657ce52428b8a65324cae5a2cc696c

BraulioV commented on ova/Libraries/elastic_functions.sh:
```diff
@@ -0,0 +1,247 @@
+
+install_logstash_6(){
+
+    yum install logstash-${2} -y
```
This line should be `yum install logstash-${ELK_VERSION}`.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/wazuh_functions.sh:
```diff
@@ -0,0 +1,71 @@
+
+set_wazuh_repository(){
+    if [ "${STATUS_PACKAGES}" == "stable" ]; then
+    # Wazuh production repository
```
Check the indentation of this line.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/wazuh_functions.sh:
```diff
@@ -0,0 +1,71 @@
+    fi
+
+    if [ "${STATUS_PACKAGES}" == "unstable" ]; then
+    # Wazuh pre-release repository
```
Check the indentation of this line.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/06b971366820c9d9cbf359ba4c3b01c427772310

BraulioV commented on ova/Libraries/wazuh_functions.sh:
```diff
@@ -0,0 +1,71 @@
+    systemctl stop wazuh-manager
+    systemctl stop wazuh-api
+
+    find ${DIRECTORY}/logs -name *.log -exec : > {} \;
```
Wazuh generates .json files as logs too. For example: `alerts.json` and `ossec.json`, so it would be interesting to delete these log files too.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/63e8681803657ce52428b8a65324cae5a2cc696c

BraulioV commented on ova/generate_ova.sh:
```diff
@@ -47,12 +47,20 @@ function build_ova() {
   local OVF_VM="wazuh${OVA_VERSION}.ovf"
   local OVA_FIXED="wazuh${OVA_VERSION}-fixed.ova"
   local OVA_VMDK="wazuh${OVA_VERSION}-disk001.vmdk"
+  local ELK_MAJOR=`echo ${ELK_VERSION}|cut -d"." -f1`
```
`local` syntax is `bash` specific syntax. If this script is executed in a host where bash isn't installed, it won't work. The best solution is to set the _shebang_ to `#!/bin/sh` and use `sh` syntax which it is available in all Unix hosts.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/63e8681803657ce52428b8a65324cae5a2cc696c

BraulioV commented on ova/Config_files/filebeat.yml:
```diff
@@ -0,0 +1,58 @@
+# Wazuh - Filebeat configuration file
```
This file isn't necessary. This file can be downloaded from https://github.com/wazuh/wazuh/blob/3.9/extensions/filebeat/7.x/filebeat.yml.

BraulioV commented on ova/.vagrant/rgloader/loader.rb:
```diff
@@ -0,0 +1,9 @@
+# This file loads the proper rgloader/loader.rb file that comes packaged
```
This file isn't necessary. You can create a `.gitignore` file to avoid committing anything from `.vagrant`.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/2629f18c2d161e7cfe70b71d39522548b5459f55

BraulioV commented on ova/generate_ova.sh:
```diff
@@ -1,4 +1,5 @@
-#!/bin/bash
+#!/bin/sh
+set -exf
```
This isn't really necessary.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/2629f18c2d161e7cfe70b71d39522548b5459f55

BraulioV commented on ova/generate_ova.sh:
```diff
@@ -19,9 +20,9 @@ scriptpath=$(
+help () {
 
-  OPTIONS:
+  echo "OPTIONS:"
```
Can you change this line to these lines: https://github.com/wazuh/wazuh-packages/blob/master/debs/generate_debian_package.sh#L110-L113. This will keep consistency between the help messages in all the different scripts.

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/2629f18c2d161e7cfe70b71d39522548b5459f55

BraulioV commented on ova/generate_ova.sh:
```diff
@@ -56,8 +57,8 @@ function build_ova() {
   if [ ${ELK_MAJOR} -eq 7 ]; then
-      curl -so Config_files/filebeat.yml https://raw.githubusercontent.com/wazuh/wazuh/v${WAZUH_VERSION}/extensions/filebeat/7.x/filebeat.yml
-      sed -i "s|#pipeline: geoip|pipeline: geoip|" Config_files/filebeat.yml
+      curl -so config_files/filebeat.yml https://raw.githubusercontent.com/wazuh/wazuh/v${WAZUH_VERSION}/extensions/filebeat/7.x/filebeat.yml
```
Be careful with this line. You are downloading the `filebeat.yml` file in `config_files` and this directory doesn't exist. Did you mean `Config_files` ?

DFolchA replied:
Done in https://github.com/wazuh/wazuh-packages/pull/223/commits/2629f18c2d161e7cfe70b71d39522548b5459f55

DFolchA commented:
Hello team,

By recommendation of @BraulioV we have decided to make the generation of the ova more sensitive to failure to do this I have added the set -e option to the provision script so it exits when something returns a value different to 0 and the option -f to all the curl orders so they don't always return 0 .

Additionally, I modified the clean function of the generate_ova script and now the script calls it if it fails at some critical points such as if the vagrant up fails to provision.

To test these changes I have generated the following OVAs:
- 3.9.0-6.7.2
- 3.9.3-7.2.0

Regards
